### ISSUE TITLE
Untracked transaction for multiple databases in System Tests

### ISSUE THREAD

**Issue by ngan** [palkan/isolator#78]
Our environment is as such:
- Rails 7.1
- RSpec
- Puma
- Multiple databases

We're noticing `Trying to finalize an untracked transaction` warnings sometimes in our system specs and were able to figure out the problem. Let's say you have 2 databases (A and B). When you start a system spec it could start with only 1 connection to A. Puma then spins up in a separate thread and starts serving requests. Keep in mind that Rails system tests makes it so that the connection pool is shared between the test suite and the Puma server. In serving the request, a connection to database B is made. Since `ThreadStateProxy` is thread-aware, it stores a `state` per thread. The connection increment is happening in the Puma thread. When the system test finishes and transactions are rolled back, the transaction in database B counter goes from 0 to -1 because according to the test suite thread, the transaction/connection was never made.

I'm still thinking about how to solve this, but I wanted to file this issue in the meantime.

**Comment by tubaxenor** [comment]
Digging a little further, then untracked connection enters https://github.com/palkan/isolator/blob/master/lib/isolator/orm_adapters/active_support_transaction_subscriber.rb#L23 which does not invoke a `Isolator.incr_transactions!(connection_id)` and causing the finalize mismatch.

**Comment by palkan** [comment]
> The connection increment is happening in the Puma thread

It also must happen in the test thread (before any browser interaction), since every test is wrapped into a transaction.

> When you start a system spec it could start with only 1 connection to A

How do you implement multiple databases? If using Active Record built-in support, a test transaction is open for every database from the configuration (so, both A and B transactions must first be open in the test thread)?

**Comment by ngan** [comment]
> a test transaction is open for every database from the configuration

I don't think this is true. Connections are lazily established in all environments. At the beginning of tests, it will only start transactions for open connections:
https://github.com/rails/rails/blob/main/activerecord/lib/active_record/test_fixtures.rb#L118-L119

> How do you implement multiple databases? If using Active Record built-in support

We're using the built-in support.
